File: LeetCode/cpp/653_Two_Sum_IV_-_Input_is_a_BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
	void check(TreeNode* root, int k, set<int>& st, bool& flag) {
		if (root == nullptr) {
			return;
		}

		if (st.find(root->val) == st.end()) {
			st.insert(k - root->val);

			check(root->left, k, st, flag);
			check(root->right, k, st, flag);
		}
		else {
			flag = true;
			return;
		}
	}

	bool findTarget(TreeNode* root, int k) {
		set<int> st;
		bool flag = false;

		check(root, k, st, flag);

		if (flag)
			return true;
		return false;
	}
};

File: LeetCode/java/653_Two_Sum_IV_-_Input_is_a_BST.java
import javax.swing.tree.TreeNode;
import java.util.HashSet;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean check(TreeNode root, int k, HashSet<Integer> hs) {
        if (root == null) {
            return false;
        }

        if (hs.contains(root.val)) {
            return true;
        }

        hs.add(k - root.val);
        return check(root.left, k, hs) || check(root.right, k, hs);
    }

    public boolean findTarget(TreeNode root, int k) {
        HashSet<Integer> hs = new HashSet<>();

        return check(root, k, hs);
    }
}